### DISCUSSION TITLE
Is it normal for beam.smp to consume nearly 50% of RAM on a 4GB VM?

### DISCUSSION THREAD

**Discussion by jrothlis** [teslamate-org/teslamate#3045] [Q&A] [answered]
I recently updated the VM my TeslaMate is running on (PhotonOS v3) and when rebooting, the VM quickly ran out of resources and hung. So I restarted it and watched htop, and after seeing it crash again, I increased the VM's RAM from 2GB to 4GB, and now it's stable. Basically there are 19 "instances" (I don't know what the correct term is) of beam.smp consuming collectively 43.9% of the 4GB of the VM. This definitely wasn't the case before as it run for 3 years with 2GB of RAM very happily.

I don't begrudge the extra RAM, but wanted to check if this is known or expected behaviour? To be clear TeslaMate itself (running under docker-compose) was already at the latest version, I just ran the underlying OS updates and rebooted the VM. So there seems to be a new interaction that has triggered this behaviour.

<img width="1083" alt="Screenshot 2023-01-07 at 11 11 22" src="https://user-images.githubusercontent.com/17953/211147302-756c724a-e6d8-49b9-8400-5cf1b1145e72.png">

**Comment by Bomfunk** [comment]
Mine consumes almost exactly the same amount of RAM. I also wish to know if this is normal, and if there is a way to reduce the usage?

**Reply by vrykolas654** [reply]
Not a solution (just a data point): I'm running Teslamate inside Docker on a Synology and all 4 docker images (together) only use ~300-400MB RAM in normal usage with 1 vehicle.

**Reply by Bomfunk** [reply]
I think it was like that in the beginning for me as well, but it grows over time. The restarts don't fix this, so I suppose this has something to do with the DB size.

**Comment by Bomfunk** [comment]
Since my last post, the beam.smp process went from 1.7 GB to 2.2 GB. Seriously, is there any way to reduce the consumption?

**Reply by dkdc** [reply]
Exactly the same for me. I cannot run teslamate any longer after the update as memory on this kvm is exhausted. Please fix.

**Comment by inigoml** [comment]
Hi everbody.

Assuming you are using docker-compose, you can easily limit resources per docker with a simple configuration applied at docker-compose.yaml. For example, in order to limit teslamate container to 200Mbs

![image](https://github.com/teslamate-org/teslamate/assets/7486708/d706da3a-1302-41d3-b656-2a3cf1f24edd)


The same can be applied to database and grafana containers. Mosquitto does not consume almost any resource, so no need to insert a hard limit.

**Reply by Bomfunk** [reply]
Thank you for the tip. However, if I do that, teslamate ends up being in a restart loop as it's being killed.

**Comment by dyxyl** [comment]
Which version(s) of erts are causing the problems (see the path containing `beam.smp`)?  @jrothlis is using erts-12.3.2.2, which appears to 2 years old now according to https://github.com/erlang/otp/releases/tag/OTP-24.3.4.2).  The latest pre-built TeslaMate docker container is using erts-14.2.1.

If you are building your own TeslaMate, then check you have a recent version of erlang/elixir.

**Reply by jrothlis** [reply]
Whatever version comes with TeslaMate. Why on earth would you assume I am building my own?

**Reply by jrothlis** [reply]
![IMG_5799](https://github.com/teslamate-org/teslamate/assets/17953/0575aa64-9bd1-4d07-8b16-ac932e9e3370)

**Comment by Saghub01** [comment]
Hello, same problem here. I am running Teslamate (1.28.3 on docker) for 1 year on a RPI3. Beam.smp now consume 21% of the RAM. Didn't notice this before.

![BeamRPI3](https://github.com/teslamate-org/teslamate/assets/162872173/02717025-8e7c-4d70-b7c9-0119e063b7d6)

**Reply by ngardiner** [reply]
I don't think that's the same thing, your rpi3 has 1GB of RAM and teslamate is consuming 21% of that or 200MB. The OP was posting about 50% utilisation on a 4GB VM (2GB) which is significantly more and which does seem high comparitively.

Mine's using 3.4% but I am running it on an 8GB pi, so if I was to extrapolate that out it would be 27.2% relative to your 1GB or about 270MB.

**Reply by Saghub01** [reply]
Ok understood, thank you. I keep reading this thread.

**Comment by inigoml** [comment]
From my installations.

First one, installed in a local NAS. 16GB RAM. No problem at all since plenty RAM available.
![image](https://github.com/teslamate-org/teslamate/assets/7486708/bd35a46d-3855-4444-88e4-4d55a068eace)

Second one, from my VM at GCP. 1GB RAM
![image](https://github.com/teslamate-org/teslamate/assets/7486708/e2cecfe5-6c4f-4425-9c6b-65bf56221436)

Take into consideration that RES memory is what really matters. That is, real memory used.

In my understanding, process RAM usage is proportional to total RAM. In the second scenario, docker memory is constrained by configuration to 200MB.

Take into consideration that BEAM is Erlan Virtual Machine and so it would try to allocate some resources in order to run same way JVM does. So, in my understanding, we could limit BEAM processs in order to constrain resources used.

**Reply by jrothlis** [reply]
I don’t understand how different installs can have such wildly different memory consumption. I’m using the official TeslaMate images and my beam.smp is currently consuming 57% of a 4GB VM, so it would be crashing in a 2GB VM. It’s fine, I have no issue letting it have 4GB, but it used to be fine until 14 months ago when I raised this issue. 

I think the real issue is that TeslaMate appears to be effectively abandoned at this point, which is sad, but understandable (life moves on).

**Reply by dyxyl** [reply]
Abandoned?  There was a release on Feb 2 this year.

If you have a spare VM, you could try starting another instance with an empty db and without connecting it to any cars to see how much RAM that needs.

**Reply by jrothlis** [reply]
I suppose...? Previous release was November (3 months prior). Maybe it's picking up steam again. Edit to say: abandoned in the sense that a bug I raised over a year ago, that appears to affect multiple people, has had zero traction.

My setup is 4 years old, surely matched by a lot of TeslaMate users. What would setting up a clean install achieve?

**Reply by dyxyl** [reply]
@inigoml Don't forget that running with swap can make the RES artificially lower than the actual memory needed. For me, changing the VM's RAM from 1GB to 4GB didn't make any difference to the VIRT or RES of the process a few minutes after it starts, so it doesn't seem to be preallocating anything.

**Reply by dyxyl** [reply]
@jrothlis It is difficult to track down bugs like this in a dynamic system and possibly it isn't causing problems for very many people (they might have more RAM or use swap, so the process size doesn't matter).

As I said, it would be interesting to try a clean install **to see how much RAM that needs**.  I've tried that on an x86 VM and it only needed 219MB of RAM.  Does it grow larger for you soon after startup or just over days and weeks?

**Reply by jrothlis** [reply]
Yeah it goes to that RAM utilisation seconds after starting. It is an x86 VM on top of ESXi 8, running PhotonOS 4, kernel 5.10.209-3.

```root@teslamate [ /var/lib/teslamate ]# uname -srm
Linux 5.10.209-3.ph4-esx x86_64
root@teslamate [ /var/lib/teslamate ]# cat /etc/os-release 
NAME="VMware Photon OS"
VERSION="4.0"
ID=photon
VERSION_ID=4.0
PRETTY_NAME="VMware Photon OS/Linux"
ANSI_COLOR="1;34"
HOME_URL="https://vmware.github.io/photon/"
BUG_REPORT_URL="https://github.com/vmware/photon/issues"
root@teslamate [ /var/lib/teslamate ]# 
```

Note that the bug originally manifested after a routine OS update (tdnf -y update --refresh) which likely included a kernel update, but wasn't a major OS update (i.e. it wasn't moving from PhotonOS v3 to v4). So it is related to how that beam.smp process interacts with Linux that changed.

"Quiet" VM after shutting down all containers (docker-compose down):
<img width="1465" alt="Screenshot 2024-03-10 at 15 16 14" src="https://github.com/teslamate-org/teslamate/assets/17953/10ebfe26-3ea9-4cc3-8052-bf5849fd07c8">

~30 seconds after starting containers (docker-compose up -d):
<img width="1466" alt="Screenshot 2024-03-10 at 15 17 33" src="https://github.com/teslamate-org/teslamate/assets/17953/a562ec76-5c4a-4586-a7a3-a4973ce29e1e">

**Reply by inigoml** [reply]
In my GCP VM I must confess that I had to create swapfile in order to deal with Teslamate. By default GCP VMs comes with no swap and so, when beam process was running, there was no available resources in order to serve requests or query database. So, I had to add 1GB swapfile. Once done, everything is running smoothly. So yes, linux OS moves memory chunks to swapfile when not needed.

**Reply by dyxyl** [reply, marked as answer]
I've tracked down the problem when using PhotonOS 4 and probably others: docker changed the max number of open files to "unlimited" (though they have subsequently reverted that change).  Unlimited is seen as 1073741824 by erts beam, which causes it to allocate a huge amount of memory on startup.

You can fix it by adding this to the teslamate section in docker-compose.yml:
```
    ulimits:
      nofile: 65536
```
and then run `docker-compose up -d`.

**Reply by Bomfunk** [reply]
That worked, thank you!

**Reply by jrothlis** [reply]
Amazing, thank you so much @dyxyl! My overall memory consumption is down to 551MB, from 2.5GB without the fix.

**Reply by nguyenleminhquan** [reply]
thank you very much, you save my months!

**Comment by dkdc** [comment]
Kudos!
[That](https://github.com/teslamate-org/teslamate/discussions/3045#discussioncomment-8757861) solved it.  Thanks.

It's down to 200k, still top 1 on that kvm-machine.

1075198 10000     20   0 1841472 **203876**  90828 S   0.0   6.8   0:04.46 beam.smp                 

Probaby there's also no point for 64k files, is there? And one could use good old unix value of 1024?

6000km of history missing now sigh ;)

**Reply by dyxyl** [reply]
Yes, 1024 should work too (check the output from lsof -p $pid or look in /proc/$pid/fd to see what it is using). I think it won't reduce the size any more though because beam effectively does max(64k,nofiles).
